art/attacks/evasion/simba.py

```diff
@@ -0,0 +1,319 @@
+        :type max_iter: `int`
+        :param epsilon: Overshoot parameter.
+        :type epsilon: `float`
+        :param order: order of pixel attacks
+        :type order: `str`
+        :param freq_dim: dimensionality of 2D frequency space (DCT).
+        :type freq_dim: `int`
+        :param stride: stride for block order (DCT).
+        :type stride: `int`
+        :param targeted: targeted attacks
+        :type targeted: `bool`
+        :param batch_size: Batch size (but, batch process unavailable in this implementation)
+        :type batch_size: `int`
+        """
+        # Save attack-specific parameters
+        super(SimBA, self).set_params(**kwargs)
+
+        if not isinstance(self.max_iter, (int, np.int)) or self.max_iter <= 0:
+            raise ValueError("The number of iterations must be a positive integer.")
+
+        if self.epsilon < 0:
+            raise ValueError("The overshoot parameter must not be negative.")
+
+        if self.batch_size != 1:
+            raise ValueError('The batch size `batch_size` has to be 1 in this implementation.')
+        
+        if not isinstance(self.stride, (int, np.int)) or self.stride <= 0:
+            raise ValueError("The `stride` value must be a positive integer.")
+        
+        if not isinstance(self.freq_dim, (int, np.int)) or self.freq_dim <= 0:
+            raise ValueError("The `freq_dim` value must be a positive integer.")
+        
+        if self.order != 'random' and self.order != 'diag':
+            raise ValueError('The order of pixel attacks has to be `random` or `diag`.')
+        
+        if self.attack != 'px' and self.attack != 'dct':
+            raise ValueError('The attack type has to be `px` or `dct`.')
+        
+        if not isinstance(self.targeted, (int)) or (self.targeted != 0 and self.targeted != 1):
+            raise ValueError('`targeted` has to be a logical value.')
+
+        return True
+
+    def _block_order(self, img_size, channels, initial_size=2, stride=1):
+        order = np.zeros((channels , img_size , img_size))
+        total_elems = channels * initial_size * initial_size
+        perm = np.random.permutation(total_elems)
+        order[:, :initial_size, :initial_size] = perm.reshape((channels, initial_size, initial_size))
+        for i in range(initial_size, img_size, stride):
+            num_elems = channels * (2 * stride * i + stride * stride)
+            perm = np.random.permutation(num_elems) + total_elems
+            num_first = channels * stride * (stride + i)
+            order[:, :(i+stride), i:(i+stride)] = perm[:num_first].reshape((channels, -1, stride))
+            order[:, i:(i+stride), :i] = perm[num_first:].reshape((channels, stride, -1))
+            total_elems += num_elems
+        return order.transpose(1,2,0).reshape(1, -1).squeeze().argsort()
+
+    # applies IDCT to each block of size block_size
+    def _block_idct(self, x, block_size=8, masked=False, ratio=0.5):
```

Comment:
Please add a docstring and typing information to this method.

Reply:
added a docstring and typing information to this method.